src/core/Akka/IO/TcpConnection.cs

```diff
@@ -829,28 +830,40 @@ public override PendingWrite DoWrite(ConnectionInfo info)
             {
                 try
                 {
-                    if (!data.IsEmpty)
+                    PendingWrite writeToChannel(ByteString data)
                     {
-                        connection.SetStatus(ConnectionStatus.Sending);
-
-                        sendArgs.SetBuffer(data);
-
-                        if (!connection.Socket.SendAsync(sendArgs))
-                            self.Tell(SocketSent.Instance);
+                        var bytesWritten = _connection.Socket.Send(data.Buffers);
+                        if (_connection.traceLogging)
+                            _connection.Log.Debug("Wrote [{0}] bytes to channel", bytesWritten);
+                        if (bytesWritten < data.Count)
+                        {
+                            // we weren't able to write all bytes from the buffer, so we need to try again later
+                            return writeToChannel(data.Slice(bytesWritten));
+                        }
+                        else // finished writing
+                        {
+                            if(Ack != NoAck.Instance) Commander.Tell(Ack);
+                            Release();
+                            return _connection.CreatePendingWrite(Commander, _tail, info);
+                        }
                     }
 
-                    var ack = Ack == NoAck.Instance ? null : Tuple.Create(Commander, Ack);
-                    connection.SetPendingAcknowledgement(ack);
-                    return connection.CreatePendingWrite(Commander, tail, info);
+                    return writeToChannel(_buffer);
+
                 }
                 catch (SocketException e)
                 {
-                    connection.HandleError(info.Handler, e);
+                    _connection.HandleError(info.Handler, e);
                     return this;
                 }
             }
 
             public override void Release() { }
```

Comment:
Really should find a way to use buffer pooling here the same way the JVM does, unless we bolt that onto `SocketAsyncEventArgs` pooling.